### PR TITLE
Postgres unload

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
This adds support for using the COPY TO statement in Postgres to unload a file.  Up until this PR, we only supported exporting using SELECT into a DatafRame.

### CODE REVIEW

File: records_mover/db/postgres/copy_options/common.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
 from records_mover.utils import quiet_remove
 from records_mover.records.hints import cant_handle_hint
 from records_mover.records.types import RecordsHints
-from typing import Set, Optional, Tuple
-from .date_input_style import DateInputStyle, determine_date_input_style
+from typing import Set, Tuple, Union, Literal
```
Comment (bluelabsbutler) [inline records_mover/db/postgres/copy_options/common.py:4]: F401 'typing.Tuple' imported but unused

File: records_mover/db/postgres/copy_options/date_input_style.py [hunk 2]
```diff
@@ -1,42 +1,22 @@
...
-    date_input_style: Optional[DateInputStyle] = None
-
-    def upgrade_date_input_style(style: DateInputStyle, hint_name: str) -> None:
-        nonlocal date_input_style
-        if date_input_style not in (None, style):
+from typing import Optional, Set
+from .types import DateOrderStyle
+
+
+def determine_input_date_order_style(unhandled_hints: Set[str],
```
Comment (bluelabsbutler) [inline records_mover/db/postgres/copy_options/date_input_style.py:8]: C901 'determine_input_date_order_style' is too complex (16)

File: records_mover/db/postgres/copy_options/text.py [hunk 3]
```diff
@@ -1,16 +1,20 @@
 from records_mover.utils import quiet_remove
 from records_mover.records.hints import cant_handle_hint
 from records_mover.records.types import RecordsHints
-from typing import Set, Optional, Tuple
-from .date_input_style import DateInputStyle
+from typing import Set, Optional, Tuple, Union, Literal
```
Comment (bluelabsbutler) [inline records_mover/db/postgres/copy_options/text.py:4]: F401 'typing.Optional' imported but unused

Comment (bluelabsbutler) [inline records_mover/db/postgres/copy_options/text.py:4]: F401 'typing.Tuple' imported but unused

File: records_mover/db/postgres/copy_options/__init__.py [hunk 4]
```diff
@@ -3,10 +3,14 @@
...
 import logging
 from typing import Set, Tuple, Optional
-from .date_input_style import DateInputStyle
+from .date_input_style import determine_input_date_order_style
+from .date_output_style import determine_output_date_order_style
 from .csv import postgres_copy_options_csv
 from .text import postgres_copy_options_text
-from .types import PostgresCopyOptions
+from .types import PostgresCopyOptions, DateOrderStyle, DateOutputStyle
+# TODO: Is this even needed?
```
Comment (bluelabsbutler) [inline records_mover/db/postgres/copy_options/__init__.py:11]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/db/postgres/copy_options/csv.py [hunk 5]
```diff
@@ -1,16 +1,20 @@
...
 from records_mover.records.types import RecordsHints
-from typing import Set, Optional, Tuple
-from .date_input_style import DateInputStyle
+from typing import Set, Union, Literal
 from .common import postgres_copy_options_common
+from .mode import CopyOptionsMode
 from .types import PostgresCopyOptions
 
 
+# TODO: Audit this file
```
Comment (bluelabsbutler) [inline records_mover/db/postgres/copy_options/csv.py:10]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/db/postgres/copy_options/text.py [hunk 6]
```diff
@@ -1,16 +1,20 @@
...
 from .types import PostgresCopyOptions
+from .mode import CopyOptionsMode
 from .common import postgres_copy_options_common
 
 
+# TODO: Audit this file
```
Comment (bluelabsbutler) [inline records_mover/db/postgres/copy_options/text.py:10]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/db/postgres/unloader.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                        fileobj,
+                        conn,
+                        **postgres_options)
+
+        logger.info('Copy complete')
+        directory.save_preliminary_manifest()
+
+    def known_supported_records_formats_for_unload(self) -> List[BaseRecordsFormat]:
+        return [
+            # TODO update these validation instructions and validate each
```
Comment (bluelabsbutler) [inline records_mover/db/postgres/unloader.py:72]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: .circleci/config.yml [hunk 8]
```diff
@@ -358,8 +358,8 @@ workflows:
             mkdir -p test-reports/itest
             cd tests/integration/records/single_db
             with-db dockerized-postgres nosetests --with-xunit --xunit-file=../../../../test-reports/itest/junit.xml .
-          requires:
-            - redshift-itest
+#          requires:  # TODO
```
Comment (bluelabsbutler) [inline .circleci/config.yml:361]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: .circleci/config.yml [hunk 9]
```diff
@@ -485,8 +485,8 @@ workflows:
             mkdir -p test-reports/itest
             cd tests/integration
             python3 -m records.multi_db.test_records_table2table
-          requires:
-            - redshift-itest
+#          requires:  # TODO
```
Comment (bluelabsbutler) [inline .circleci/config.yml:488]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: tests/unit/db/postgres/test_unloader.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        mock_table_obj = mock_Table.return_value
+        mock_Table.assert_called_with(mock_table,
+                                      ANY,
+                                      schema=mock_schema,
+                                      autoload=True,
+                                      autoload_with=self.mock_db)
+        mock_conn = self.mock_db.engine.begin.return_value.__enter__.return_value
+        mock_quote_value.assert_called_with(mock_conn, 'DATE_OUTPUT_STYLE, DATE_ORDER_STYLE')
+        mock_conn.execute.assert_called_with('SET LOCAL DateStyle = ABC')
+        mock_fileobj = mock_directory.loc.file_in_this_directory.return_value.open.return_value.__enter__.return_value
```
Comment (bluelabsbutler) [inline tests/unit/db/postgres/test_unloader.py:63]: E501 line too long (118 > 100 characters)

File: tests/unit/db/postgres/test_unloader.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        mock_table_obj = mock_Table.return_value
+        mock_Table.assert_called_with(mock_table,
+                                      ANY,
+                                      schema=mock_schema,
+                                      autoload=True,
+                                      autoload_with=self.mock_db)
+        mock_conn = self.mock_db.engine.begin.return_value.__enter__.return_value
+        mock_quote_value.assert_called_with(mock_conn, 'DATE_OUTPUT_STYLE, MDY')
+        mock_conn.execute.assert_called_with('SET LOCAL DateStyle = ABC')
+        mock_fileobj = mock_directory.loc.file_in_this_directory.return_value.open.return_value.__enter__.return_value
```
Comment (bluelabsbutler) [inline tests/unit/db/postgres/test_unloader.py:119]: E501 line too long (118 > 100 characters)

File: records_mover/db/postgres/copy_options/csv.py [hunk 12]
```diff
@@ -70,12 +71,21 @@ def postgres_copy_options_csv(unhandled_hints: Set[str],
...
-
-    if hints['quoting'] != 'minimal':
-        cant_handle_hint(fail_if_cant_handle_hint, 'quoting', hints)
+    if mode is CopyOptionsMode.LOADING:
+        if hints['quoting'] != 'minimal':
+            cant_handle_hint(fail_if_cant_handle_hint, 'quoting', hints)
+        else:
+            quiet_remove(unhandled_hints, 'quoting')
+    elif mode is CopyOptionsMode.UNLOADING:
+        # TODO: Are we sure this doesn't support minimal by default instead of None by default?
```
Comment (bluelabsbutler) [inline records_mover/db/postgres/copy_options/csv.py:80]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/db/driver.py [hunk 13]
```diff
@@ -56,7 +56,7 @@ def unload(self,
                schema: str,
                table: str,
                unload_plan: RecordsUnloadPlan,
-               directory: RecordsDirectory) -> int:
+               directory: RecordsDirectory) -> Optional[int]:
```
Comment (vinceatbluelabs) [inline records_mover/db/driver.py:59]: Postgres doesn't supply the number of rows exported via its COPY TO statement.  Boo.

File: records_mover/db/postgres/copy_options/__init__.py [hunk 14]
```diff
@@ -34,20 +37,59 @@ def needs_csv_format(hints: RecordsHints) -> bool:
...
+    if needs_csv_format(hints):
+        copy_options = postgres_copy_options_csv(unhandled_hints,
+                                                 hints,
+                                                 fail_if_cant_handle_hint,
+                                                 CopyOptionsMode.UNLOADING)
+    else:
+        copy_options = postgres_copy_options_text(unhandled_hints,
+                                                  hints,
+                                                  fail_if_cant_handle_hint,
+                                                  CopyOptionsMode.UNLOADING)
```
Comment (vinceatbluelabs) [inline records_mover/db/postgres/copy_options/__init__.py:58]: Postgres uses the same fundamental COPY statement for both loading and unloading, but there is some nuance for some of the bits, so I'm passing in a flag here.

File: records_mover/db/postgres/copy_options/__init__.py [hunk 15]
```diff
@@ -34,20 +37,59 @@ def needs_csv_format(hints: RecordsHints) -> bool:
...
+
+    date_output_style, date_order_style =\
+        determine_output_date_order_style(unhandled_hints,
+                                          hints,
+                                          fail_if_cant_handle_hint)
```
Comment (vinceatbluelabs) [inline records_mover/db/postgres/copy_options/__init__.py:63]: The semantics of the date-related commands is fundamentally different, though - Postgres is very limited on date output, but very forgiving on date input, so I broke the handling of them out into two separate input/output functions.

File: records_mover/db/postgres/copy_options/date_input_style.py [hunk 16]
```diff
@@ -1,42 +1,22 @@
...
-    #  synonyms for MDY. See Section 8.5 for more information. The
-    #  built-in default is ISO, MDY, but initdb will initialize the
-    #  configuration file with a setting that corresponds to the
-    #  behavior of the chosen lc_time locale.
-    DateInputStyle = Union[Literal["DMY"], Literal["MDY"]]
-else:
-    DateInputStyle = str
-
-
-def determine_date_input_style(unhandled_hints: Set[str],
```
Comment (vinceatbluelabs) [inline records_mover/db/postgres/copy_options/date_input_style.py:28]: I got the terminology slightly wrong - this is a date order style in postgres terminology.  It's only really relevant on import.

File: records_mover/db/postgres/loader.py [hunk 17]
```diff
@@ -119,8 +119,9 @@ def can_load_this_format(self, source_records_format: BaseRecordsFormat) -> bool
 
     def known_supported_records_formats_for_load(self) -> List[BaseRecordsFormat]:
         return [
-            # To validate that these load without, watch logging while
-            # running and verify 'dataframe' doesn't appear:
+            # To validate that these load without pandas, watch
+            # logging while running and verify 'dataframe' doesn't
+            # appear:
```
Comment (vinceatbluelabs) [inline records_mover/db/postgres/loader.py:124]: I accidentally a word.

File: records_mover/db/postgres/copy_options/date_output_style.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+from records_mover.records.types import RecordsHints
+from records_mover.records.hints import cant_handle_hint
+from typing import Set, Tuple, Optional
+from .types import DateOrderStyle, DateOutputStyle
+
+
+def determine_date_output_style(unhandled_hints: Set[str],
+                                hints: RecordsHints,
+                                fail_if_cant_handle_hint: bool) -> \
+        Tuple[DateOutputStyle, Optional[DateOrderStyle]]:
```
Comment (vinceatbluelabs) [inline records_mover/db/postgres/copy_options/date_output_style.py:11]: The semantics of the date-related commands is fundamentally different, though - Postgres is very limited on date output, but very forgiving on date input, so I broke the handling of them out into two separate input/output functions.

**Comment by vinceatbluelabs** [comment]
Hi @cwegrzyn!  I just updated this back from master.  It's ready for review when you have a chance - it's not blocking anything big, though, so no huge rush.

**Comment by vinceatbluelabs** [comment]
Hi @cwegrzyn!  If you find the time to review, this one's starting to heat up.  I'm realizing as I start to dive in on documenting writing the bulk loading parts of the database driver along with the MySQL loading support that it'd be good to refactor the rather large DBDriver interface.  I'm thinking it'll be smart to get this one merged before I go too far in that direction.

**Comment by vinceatbluelabs** [comment]
@cwegrzyn: I know you're super busy right now--I'm going to go ahead and merge, but feel free to mark this up and I'll backlog any improvements.  Thanks as always for what time you can offer on these--the feedback is always enlightening!

**Pull request merged**